File: gateway/ts/src/core/stream.ts
import { type GatewayError, GatewayFailure, invalid } from "./protocol.js";

export interface StreamPolicy {
  maxChunkBytes: number;
  queueItems: number;
  queueBytes: number;
  maxOwnerStreams: number;
  maxCallerStreams: number;
  openTimeoutMs: number;
  producerIdleMs: number;
  consumerIdleMs: number;
  totalMs: number;
}
export const DEFAULT_STREAM_POLICY: Readonly<StreamPolicy> = Object.freeze({
  maxChunkBytes: 1024 * 1024,
  queueItems: 16,
  queueBytes: 4 * 1024 * 1024,
  maxOwnerStreams: 128,
  maxCallerStreams: 32,
  openTimeoutMs: 10_000,
  producerIdleMs: 60_000,
  consumerIdleMs: 60_000,
  totalMs: 0,
});
export function streamPolicy(options: Partial<StreamPolicy> = {}): StreamPolicy {
  const policy = { ...DEFAULT_STREAM_POLICY, ...options };
  for (const [name, value] of Object.entries(policy)) {
    const max =
      name === "maxChunkBytes" || name === "queueBytes"
        ? 64 * 1024 * 1024
        : name.endsWith("Streams") || name === "queueItems"
          ? 4096
          : 2_147_483_647;
    if (!Number.isSafeInteger(value) || value < (name === "totalMs" ? 0 : 1) || value > max)
      invalid(`invalid stream policy: ${name}`);
  }
  return policy;
}
export interface ResponseStream<T> extends AsyncIterableIterator<T> {
  return(): Promise<IteratorResult<T>>;
  readonly policy: Readonly<StreamPolicy>;
  readonly closed: Promise<void>;
  cancel(): Promise<void>;
}
export interface StreamOptions {
  signal?: AbortSignal;
}
export type ByteSource = AsyncIterable<Uint8Array>;
export const streamError = (code: GatewayError["code"], message: string) => new GatewayFailure({ code, message });

/** Registers cancellation before opening; no source.next() until the consumer requests it. */
export async function openStream(
  factory: (signal: AbortSignal) => Promise<ByteSource> | ByteSource,
  policy: StreamPolicy,
  options: StreamOptions = {},
  released: () => void = () => {},
): Promise<ResponseStream<Uint8Array>> {
  let settled!: () => void;
  const closed = new Promise<void>((resolve) => {
    settled = resolve;
  });
  let activeWork: Promise<unknown> = Promise.resolve();
  const controller = new AbortController();
  let source: AsyncIterator<Uint8Array> | undefined;
  let terminal: Error | "end" | undefined;
  let busy = false;
  let timer: ReturnType<typeof setTimeout> | undefined;
  let total: ReturnType<typeof setTimeout> | undefined;
  let rejectPending: ((error: Error) => void) | undefined;
  let cleanup: Promise<void> = Promise.resolve();
  let actualCleanup: Promise<unknown> = Promise.resolve();
  const dispose = () => {
    if (!source?.return) return;
    const iterator = source;
    source = undefined;
    actualCleanup = Promise.resolve().then(() => iterator.return?.());
    void actualCleanup.catch(() => {});
  };
  const finish = (reason: Error | "end") => {
    if (terminal) return;
    terminal = reason;
    clearTimeout(timer);
    clearTimeout(total);
    options.signal?.removeEventListener("abort", abort);
    controller.abort(reason);
    rejectPending?.(reason === "end" ? streamError("CANCELLED", "stream ended") : reason);
    dispose();
    settled();
    const completed = activeWork.catch(() => {}).then(() => actualCleanup);
    cleanup = new Promise<void>((resolve, reject) => {
      const timeout = setTimeout(() => reject(streamError("TIMEOUT", "producer cleanup timed out")), 1000);
      completed.then(
        () => {
          clearTimeout(timeout);
          resolve();
        },
        (error) => {
          clearTimeout(timeout);
          reject(error);
        },
      );
    });
    void cleanup.catch(() => {});
    void completed.catch(() => {}).then(released);
  };
  const abort = () =>
    finish(
      options.signal?.reason instanceof GatewayFailure
        ? options.signal.reason
        : streamError("CANCELLED", "stream cancelled"),
    );
  const deadline = (ms: number, phase: string) => {
    clearTimeout(timer);
    timer = setTimeout(() => finish(streamError("TIMEOUT", `${phase} timed out`)), ms);
  };
  const race = <T>(work: Promise<T>): Promise<T> =>
    new Promise((resolve, reject) => {
      rejectPending = reject;
      work.then(resolve, reject);
      if (terminal) reject(terminal);
    });
  options.signal?.addEventListener("abort", abort, { once: true });
  if (options.signal?.aborted) abort();
  if (terminal) throw terminal;
  deadline(policy.openTimeoutMs, "stream open");
  try {
    activeWork = Promise.resolve()
      .then(() => {
        if (terminal) throw terminal;
        return factory(controller.signal);
      })
      .then((iterable) => {
        source = iterable[Symbol.asyncIterator]();
        if (terminal) dispose();
      });
    await race(activeWork);
  } catch (error) {
    finish(error instanceof Error ? error : streamError("HANDLER_ERROR", "stream open failed"));
    throw terminal;
  }
  rejectPending = undefined;
  deadline(policy.consumerIdleMs, "consumer idle");
  if (policy.totalMs)
    total = setTimeout(() => finish(streamError("TIMEOUT", "stream total timed out")), policy.totalMs);
  const stream: ResponseStream<Uint8Array> = {
    policy: Object.freeze({ ...policy }),
    closed,
    [Symbol.asyncIterator]() {
      return this;
    },
    async next() {
      if (terminal === "end") return { done: true, value: undefined };
      if (terminal) throw terminal;
      if (busy) throw streamError("CONCURRENCY_FULL", "one pending next per stream");
      busy = true;
      deadline(policy.producerIdleMs, "producer idle");
      try {
        const iterator = source;
        if (!iterator) throw streamError("HANDLER_ERROR", "missing stream source");
        const work = Promise.resolve().then(() => iterator.next());
        activeWork = work;
        const item = await race(work);
        rejectPending = undefined;
        if (terminal) throw terminal;
        if (item.done) {
          finish("end");
          return { done: true, value: undefined };
        }
        if (!(item.value instanceof Uint8Array)) throw streamError("HANDLER_ERROR", "chunk must be PB bytes");
        if (item.value.byteLength > policy.maxChunkBytes) throw streamError("RESOURCE_EXHAUSTED", "chunk too large");
        deadline(policy.consumerIdleMs, "consumer idle");
        return { done: false, value: item.value };
      } catch (error) {
        finish(error instanceof Error ? error : streamError("HANDLER_ERROR", "stream failed"));
        throw terminal;
      } finally {
        busy = false;
        rejectPending = undefined;
      }
    },
    async cancel() {
      abort();
      await cleanup;
    },
    async return() {
      await this.cancel();
      return { done: true, value: undefined };
    },
  };
  return stream;
}

/** Active sources must await send. Concurrent/unpausable producers fail instead of accumulating waiters. */
export function boundedByteQueue(policy: StreamPolicy, signal: AbortSignal) {
  const queue: Uint8Array[] = [];
  let bytes = 0;
  let ended = false;
  let error: Error | undefined;
  let sending = false;
  let wake: (() => void) | undefined;
  let room: (() => void) | undefined;
  const notify = () => {
    wake?.();
    wake = undefined;
    room?.();
    room = undefined;
  };
  const fail = (reason: Error) => {
    error ??= reason;
    ended = true;
    queue.length = 0;
    bytes = 0;
    notify();
  };
  const abort = () => fail(streamError("CANCELLED", "queue cancelled"));
  signal.addEventListener("abort", abort, { once: true });
  if (signal.aborted) abort();
  const source: ByteSource = {
    async *[Symbol.asyncIterator]() {
      try {
        for (;;) {
          if (error) throw error;
          const value = queue.shift();
          if (value) {
            bytes -= value.byteLength;
            room?.();
            room = undefined;
            yield value;
          } else if (ended) return;
          else
            await new Promise<void>((resolve) => {
              wake = resolve;
            });
        }
      } finally {
        signal.removeEventListener("abort", abort);
        abort();
      }
    },
  };
  return {
    source,
    async send(value: Uint8Array) {
      if (sending) {
        const reason = streamError("RESOURCE_EXHAUSTED", "producer did not await send");
        fail(reason);
        throw reason;
      }
      if (value.byteLength > Math.min(policy.maxChunkBytes, policy.queueBytes)) {
        const reason = streamError("RESOURCE_EXHAUSTED", "chunk too large");
        fail(reason);
        throw reason;
      }
      sending = true;
      try {
        while (!ended && (queue.length >= policy.queueItems || bytes + value.byteLength > policy.queueBytes))
          await new Promise<void>((resolve) => {
            room = resolve;
          });
        if (error) throw error;
        if (ended) throw streamError("CANCELLED", "queue closed");
        queue.push(value);
        bytes += value.byteLength;
        wake?.();
        wake = undefined;
      } finally {
        sending = false;
      }
    },
    end() {
      ended = true;
      notify();
    },
    fail,
    usage: () => ({ items: queue.length, bytes }),
  };
}

// Version 1 binary frame: end=0; chunk=1 + big-endian u32 sequence + original PB bytes.
export function encodeFrame(seq: number, item: IteratorResult<Uint8Array>): Uint8Array {
  if (item.done) return Uint8Array.of(0);
  if (seq > 0xffffffff) throw streamError("RESOURCE_EXHAUSTED", "stream sequence exhausted");
  const bytes = new Uint8Array(5 + item.value.length);
  bytes[0] = 1;
  new DataView(bytes.buffer).setUint32(1, seq);
  bytes.set(item.value, 5);
  return bytes;
}
export function decodeFrame(bytes: Uint8Array, seq: number): IteratorResult<Uint8Array> {
  if (bytes.length === 1 && bytes[0] === 0) return { done: true, value: undefined };
  if (
    bytes.length < 5 ||
    bytes[0] !== 1 ||
    new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength).getUint32(1) !== seq
  )
    throw streamError("INCOMPATIBLE", "invalid stream sequence or frame");
  return { done: false, value: bytes.subarray(5) };
}

File: gateway/rust/src/stream.rs
//! Pull-only response streams. The actor polls the source exactly once per next.
use crate::invoke::Owner;
use crate::protocol::{ErrorCode, GatewayError};
use futures_util::{FutureExt, Stream, StreamExt};
use serde::{Deserialize, Serialize};
use std::future::Future;
use std::pin::Pin;
use std::sync::{
    Arc, Weak,
    atomic::{AtomicBool, Ordering},
};
use std::task::{Context, Poll};
use std::time::Duration;
use tokio::sync::{mpsc, oneshot, watch};

pub type ByteStream = Pin<Box<dyn Stream<Item = Result<Vec<u8>, GatewayError>> + Send>>;
pub type OpenFuture = Pin<Box<dyn Future<Output = Result<ByteStream, GatewayError>> + Send>>;
pub type StreamHandler = Arc<dyn Fn(Vec<u8>, crate::Client) -> OpenFuture + Send + Sync>;
pub struct RemoteOpen {
    pub source: ByteStream,
    pub policy: StreamPolicy,
}
pub type RemoteOpenFuture = Pin<Box<dyn Future<Output = Result<RemoteOpen, GatewayError>> + Send>>;
pub type RemoteStream = Arc<dyn Fn(crate::Route, Vec<u8>, crate::CallContext) -> RemoteOpenFuture + Send + Sync>;

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", default)]
pub struct StreamPolicy {
    pub max_chunk_bytes: usize,
    pub queue_items: usize,
    pub queue_bytes: usize,
    pub max_owner_streams: usize,
    pub max_caller_streams: usize,
    pub open_timeout_ms: u64,
    pub producer_idle_ms: u64,
    pub consumer_idle_ms: u64,
    pub total_ms: u64,
}
impl Default for StreamPolicy {
    fn default() -> Self {
        Self {
            max_chunk_bytes: 1024 * 1024,
            queue_items: 16,
            queue_bytes: 4 * 1024 * 1024,
            max_owner_streams: 128,
            max_caller_streams: 32,
            open_timeout_ms: 10_000,
            producer_idle_ms: 60_000,
            consumer_idle_ms: 60_000,
            total_ms: 0,
        }
    }
}
impl StreamPolicy {
    pub fn validate(&self) -> Result<(), GatewayError> {
        for (value, max) in [
            (self.max_chunk_bytes as u64, 64 * 1024 * 1024),
            (self.queue_bytes as u64, 64 * 1024 * 1024),
            (self.queue_items as u64, 4096),
            (self.max_owner_streams as u64, 4096),
            (self.max_caller_streams as u64, 4096),
            (self.open_timeout_ms, 2_147_483_647),
            (self.producer_idle_ms, 2_147_483_647),
            (self.consumer_idle_ms, 2_147_483_647),
        ] {
            if value == 0 || value > max {
                return Err(error(ErrorCode::InvalidArgument, "invalid stream policy"));
            }
        }
        if self.total_ms > 2_147_483_647 {
            return Err(error(ErrorCode::InvalidArgument, "invalid total timeout"));
        }
        Ok(())
    }
}
pub(crate) fn error(code: ErrorCode, message: &str) -> GatewayError {
    GatewayError::new(code, message)
}
type Item = Result<Option<Vec<u8>>, GatewayError>;
type Request = oneshot::Sender<Item>;
pub(crate) struct Life {
    pub caller: String,
    pub cancel: watch::Sender<bool>,
    pub(crate) terminal: watch::Sender<Option<Result<(), GatewayError>>>,
    busy: AtomicBool,
    requests: mpsc::Sender<Request>,
}
impl Drop for Life {
    fn drop(&mut self) {
        self.cancel.send_replace(true);
    }
}
#[derive(Clone)]
pub struct StreamHandle {
    pub(crate) life: Arc<Life>,
}
impl StreamHandle {
    pub fn is_terminal(&self) -> bool {
        self.life.terminal.borrow().is_some()
    }
    pub fn cancel_now(&self) {
        self.life.cancel.send_replace(true);
    }
    pub async fn cancel(&self) -> Result<(), GatewayError> {
        self.cancel_now();
        let mut terminal = self.life.terminal.subscribe();
        tokio::time::timeout(Duration::from_secs(1), async {
            while terminal.borrow().is_none() {
                if terminal.changed().await.is_err() {
                    break;
                }
            }
        })
        .await
        .map_err(|_| error(ErrorCode::Timeout, "producer cleanup timed out"))
    }
    pub async fn next(&self) -> Item {
        if let Some(result) = self.life.terminal.borrow().clone() {
            return result.map(|_| None);
        }
        if self.life.busy.swap(true, Ordering::AcqRel) {
            return Err(error(ErrorCode::ConcurrencyFull, "one pending next per stream"));
        }
        struct Pending<'a>(&'a AtomicBool, &'a watch::Sender<bool>, bool);
        impl Drop for Pending<'_> {
            fn drop(&mut self) {
                self.0.store(false, Ordering::Release);
                if !self.2 {
                    self.1.send_replace(true);
                }
            }
        }
        let mut pending = Pending(&self.life.busy, &self.life.cancel, false);
        let (send, recv) = oneshot::channel();
        let result = if self.life.requests.try_send(send).is_err() {
            self.life
                .terminal
                .borrow()
                .clone()
                .unwrap_or_else(|| Err(error(ErrorCode::OwnerUnavailable, "stream closed")))
                .map(|_| None)
        } else {
            recv.await
                .unwrap_or_else(|_| Err(error(ErrorCode::OwnerUnavailable, "stream closed")))
        };
        pending.2 = true;
        result
    }
}
pub(crate) struct Admission {
    pub cancel: watch::Sender<bool>,
    pub caller: String,
    pub owner: u64,
}
pub(crate) type Admissions = std::sync::Mutex<Vec<Weak<Admission>>>;

pub(crate) fn start(
    mut source: ByteStream,
    policy: StreamPolicy,
    caller: String,
    owner: Owner,
    admission: Arc<Admission>,
) -> StreamHandle {
    let (requests, mut receiver) = mpsc::channel::<Request>(1);
    let (cancel, mut cancelled) = watch::channel(false);
    let (terminal, _) = watch::channel(None);
    let life = Arc::new(Life {
        caller,
        cancel,
        terminal: terminal.clone(),
        busy: AtomicBool::new(false),
        requests,
    });
    let mut closed = owner.closed.subscribe();
    tokio::spawn(async move {
        let _admission = admission;
        let _owner = owner;
        let total = async {
            if policy.total_ms == 0 {
                std::future::pending::<()>().await;
            }
            tokio::time::sleep(Duration::from_millis(policy.total_ms)).await;
        };
        tokio::pin!(total);
        let mut pending: Option<Request> = None;
        loop {
            let work = async {
                let request = tokio::time::timeout(Duration::from_millis(policy.consumer_idle_ms), receiver.recv())
                    .await
                    .map_err(|_| error(ErrorCode::Timeout, "consumer idle timed out"))?
                    .ok_or_else(|| error(ErrorCode::Cancelled, "stream dropped"))?;
                pending = Some(request);
                let item = tokio::time::timeout(
                    Duration::from_millis(policy.producer_idle_ms),
                    std::panic::AssertUnwindSafe(source.next()).catch_unwind(),
                )
                .await
                .map_err(|_| error(ErrorCode::Timeout, "producer idle timed out"))?
                .map_err(|_| error(ErrorCode::HandlerError, "stream producer panicked"))?;
                match item {
                    None => Ok(None),
                    Some(Err(error)) => Err(error),
                    Some(Ok(bytes)) if bytes.len() > policy.max_chunk_bytes => {
                        Err(error(ErrorCode::ResourceExhausted, "chunk too large"))
                    }
                    Some(Ok(bytes)) => Ok(Some(bytes)),
                }
            };
            let item = tokio::select! {
                biased;
                _ = async { if !*cancelled.borrow() { let _ = cancelled.changed().await; } } =>
                    Err(error(ErrorCode::Cancelled, "stream cancelled")),
                _ = async { if !*closed.borrow() { let _ = closed.changed().await; } } =>
                    Err(error(ErrorCode::OwnerUnavailable, "owner closed")),
                _ = &mut total => Err(error(ErrorCode::Timeout, "stream total timed out")),
                result = work => result,
            };
            let done = !matches!(&item, Ok(Some(_)));
            let terminal_result = item.as_ref().map(|_| ()).map_err(Clone::clone);
            if done {
                drop(source);
                drop(_admission);
                terminal.send_replace(Some(terminal_result));
                if let Some(request) = pending.take() {
                    let _ = request.send(item);
                }
                return;
            }
            if let Some(request) = pending.take() {
                let _ = request.send(item);
            }
        }
    });
    StreamHandle { life }
}

/// Owned Stream adapter: dropping it releases the final handle and cancels the actor.
pub struct ResponseStream {
    handle: StreamHandle,
    pending: Option<Pin<Box<dyn Future<Output = Item> + Send>>>,
    ended: bool,
}
impl ResponseStream {
    pub fn new(handle: StreamHandle) -> Self {
        Self {
            handle,
            pending: None,
            ended: false,
        }
    }
    pub fn cancel_handle(&self) -> StreamHandle {
        self.handle.clone()
    }
    pub async fn cancel(&self) -> Result<(), GatewayError> {
        self.handle.cancel().await
    }
}
impl Stream for ResponseStream {
    type Item = Result<Vec<u8>, GatewayError>;
    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
        if self.ended {
            return Poll::Ready(None);
        }
        if self.pending.is_none() {
            let handle = self.handle.clone();
            self.pending = Some(Box::pin(async move { handle.next().await }));
        }
        match self.pending.as_mut().unwrap().as_mut().poll(cx) {
            Poll::Pending => Poll::Pending,
            Poll::Ready(item) => {
                self.pending = None;
                match item {
                    Ok(Some(bytes)) => Poll::Ready(Some(Ok(bytes))),
                    Ok(None) => {
                        self.ended = true;
                        Poll::Ready(None)
                    }
                    Err(error) => {
                        self.ended = true;
                        Poll::Ready(Some(Err(error)))
                    }
                }
            }
        }
    }
}
impl Drop for ResponseStream {
    fn drop(&mut self) {
        self.handle.cancel_now();
    }
}

pub fn encode_frame(seq: u32, item: Option<Vec<u8>>) -> Vec<u8> {
    match item {
        None => vec![0],
        Some(bytes) => {
            let mut frame = vec![1];
            frame.extend(seq.to_be_bytes());
            frame.extend(bytes);
            frame
        }
    }
}
pub fn decode_frame(bytes: Vec<u8>, seq: u32) -> Item {
    if bytes == [0] {
        return Ok(None);
    }
    if bytes.len() < 5 || bytes[0] != 1 || bytes[1..5] != seq.to_be_bytes() {
        return Err(error(ErrorCode::Incompatible, "invalid stream sequence or frame"));
    }
    Ok(Some(bytes[5..].to_vec()))
}

/// A single producer owns Sender and must await send; the queue accounts encoded PB bytes.
/// In addition to the queue, at most one chunk is held by send and one by the consumer.
pub struct ByteSender {
    sender: mpsc::Sender<(Vec<u8>, tokio::sync::OwnedSemaphorePermit)>,
    bytes: Arc<tokio::sync::Semaphore>,
    failed: watch::Sender<Option<GatewayError>>,
    max_chunk: usize,
}
impl ByteSender {
    pub async fn send(&mut self, bytes: Vec<u8>) -> Result<(), GatewayError> {
        if bytes.len() > self.max_chunk {
            let error = error(ErrorCode::ResourceExhausted, "chunk exceeds queue policy");
            self.failed.send_replace(Some(error.clone()));
            return Err(error);
        }
        let permit = tokio::select! {
            _ = self.sender.closed() => return Err(error(ErrorCode::Cancelled, "queue closed")),
            permit = self.bytes.clone().acquire_many_owned(bytes.len() as u32) =>
                permit.map_err(|_| error(ErrorCode::Cancelled, "queue closed"))?,
        };
        self.sender
            .send((bytes, permit))
            .await
            .map_err(|_| error(ErrorCode::Cancelled, "queue closed"))
    }
}
pub fn bounded_byte_queue(policy: &StreamPolicy) -> Result<(ByteSender, ByteStream), GatewayError> {
    policy.validate()?;
    let (sender, receiver) = mpsc::channel::<(Vec<u8>, tokio::sync::OwnedSemaphorePermit)>(policy.queue_items);
    let bytes = Arc::new(tokio::sync::Semaphore::new(policy.queue_bytes));
    let (failed, failure) = watch::channel(None);
    let stream = futures_util::stream::try_unfold((receiver, failure), |(mut receiver, mut failure)| async move {
        let result = tokio::select! {
            biased;
            _ = async { if failure.borrow().is_none() { let _ = failure.changed().await; } } => {
                let error = failure.borrow().clone();
                if let Some(error) = error { return Err(error); }
                receiver.recv().await
            }
            item = receiver.recv() => item,
        };
        Ok(result.map(|(bytes, permit)| {
            drop(permit);
            (bytes, (receiver, failure))
        }))
    });
    Ok((
        ByteSender {
            sender,
            bytes,
            failed,
            max_chunk: policy.max_chunk_bytes.min(policy.queue_bytes),
        },
        Box::pin(stream),
    ))
}
